src/query/storage/m3/multi_fetch_result.go

```diff
@@ -21,117 +21,151 @@
+
+	Close() error
+}
+
 // TODO: use a better seriesIterators merge here
-type multiFetchResult struct {
+type multiResult struct {
 	sync.Mutex
-	sync.Once
-	iterators        encoding.SeriesIterators
-	err              xerrors.MultiError
-	dedupeFirstAttrs storage.Attributes
-	dedupeMap        map[string]multiFetchResultSeries
-	// Need to keep track of seen iterators, otherwise they are not
-	// properly cleaned up and returned to the pool; can also leak
-	// duplciate series iterators.
-	seenIters []encoding.SeriesIterators
-	pools     encoding.IteratorPools
+	dedupeMap map[string][]multiResultSeries
+	seenIters []encoding.SeriesIterators // track known iterators to avoid leaking
+	err       xerrors.MultiError
+
+	pools encoding.IteratorPools
 }
 
-type multiFetchResultSeries struct {
-	idx   int
+func newMultiFetchResult(pools encoding.IteratorPools) multiFetchResult {
+	return &multiResult{
+		pools: pools,
+	}
+}
+
+type multiResultSeries struct {
 	attrs storage.Attributes
+	iter  encoding.SeriesIterator
 }
 
-func (r *multiFetchResult) close() error {
+func (r *multiResult) Close() error {
 	for _, iters := range r.seenIters {
 		iters.Close()
 	}
 
 	return nil
 }
 
-func (r *multiFetchResult) setPools(pools encoding.IteratorPools) {
+func (r *multiResult) FinalResult() (encoding.SeriesIterators, error) {
 	r.Lock()
-	r.Do(func() {
-		r.pools = pools
-	})
-	r.Unlock()
+	defer r.Unlock()
+	err := r.err.FinalError()
+	if err != nil {
+		return nil, err
+	}
+
+	numSeries := len(r.dedupeMap)
+	if numSeries == 0 {
+		return encoding.EmptySeriesIterators, nil
+	}
+
+	// can short-cicuit in this case
+	if len(r.seenIters) == 1 {
+		return r.seenIters[0], nil
+	}
+
+	// otherwise have to create a new seriesiters
+	iter := r.pools.MutableSeriesIterators().Get(numSeries)
+	iter.Reset(numSeries)
+
+	i := 0
+	for _, res := range r.dedupeMap {
+		if len(res) != 1 {
+			return nil, fmt.Errorf("internal error during result dedupe, expected 1 id, observed: %d", len(res))
+		}
+		iter.SetAt(i, res[0].iter)
+		i++
+	}
+
+	return iter, nil
 }
 
-func (r *multiFetchResult) add(
+func (r *multiResult) Add(
 	attrs storage.Attributes,
 	iterators encoding.SeriesIterators,
 	err error,
 ) {
 	r.Lock()
 	defer r.Unlock()
-	r.seenIters = append(r.seenIters, iterators)
 
 	if err != nil {
 		r.err = r.err.Add(err)
 		return
 	}
 
-	if r.iterators == nil {
-		r.iterators = iterators
-		r.dedupeFirstAttrs = attrs
+	r.seenIters = append(r.seenIters, iterators)
+	if !r.err.Empty() {
+		// don't need to do anything if the final result is going to be an error
 		return
 	}
 
 	iters := iterators.Iters()
-	// Need to dedupe
 	if r.dedupeMap == nil {
-		r.dedupeMap = make(map[string]multiFetchResultSeries, len(iters))
-		for idx, s := range iters {
-			r.dedupeMap[s.ID().String()] = multiFetchResultSeries{
-				idx:   idx,
-				attrs: r.dedupeFirstAttrs,
-			}
-		}
+		r.dedupeMap = make(map[string][]multiResultSeries, len(iters))
```

Comment:
Hm now we are always going to be making the dedupe map (regardless of if we only hit a single namespace)?